components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -294,4 +294,53 @@ ABIDecode(const std::vector<std::string>& types,
   return std::make_tuple(params, args);
 }
 
+absl::optional<std::vector<std::tuple<std::vector<std::string>,    // params
+                                      std::vector<std::string>>>>  // args
+ABIDecodeBalanceScannerResult(const std::vector<uint8_t>& data) {
+  // Decode first 32 bytes to get the offset of the length of the results
+  // array.
+  auto offset = GetSizeFromData(data, 0);
+  if (!offset) {
+    return absl::nullopt;
+  }
+
+  // Decode the length of the results array.
+  auto array_len = GetSizeFromData(data, *offset);
+  if (!array_len) {
+    return absl::nullopt;
+  }
+
+  // Increment the offset 32 bytes to get the offset of the first result.
+  *offset += 32;
+
+  // Decode each result in the array.
+  std::vector<std::tuple<std::vector<std::string>, std::vector<std::string>>>
+      results;
+  for (size_t i = 0; i < *array_len; i++) {
+    // Get the offset corresponding to the offset of the start of the element
+    // Should be 64 for if i = 0, 96 for if i = 1, etc.
+    auto element_offset = GetSizeFromData(data, *offset + (i * 32));
+    if (!element_offset) {
+      return absl::nullopt;
+    }
+
+    // Add the overall offset to get the offset of
+    // the start of the element
+    *element_offset += *offset;
+
+    // Take subset of data corresponding to the element
+    std::vector<uint8_t> element_data(data.begin() + *element_offset,
+                                      data.end());
+    // Decode the element.
+    auto result = ABIDecode({"bool", "bytes"}, element_data);
+    if (!result) {
+      return absl::nullopt;
+    }
+
+    results.push_back(*result);
+  }
+
+  return results;
+}
+
```

Comment:
cc @supermassive for thoughts

Reply:
cc @supermassive for thoughts

Reply:
I would suggest to reuse eth_abi_utils.h for both encoding and decoding but after looking a bit deeply into required format it seems there is no drop in replacement.

Reply:
Actually the format of result looks very similary to what is parsed by ExtractStringArrayFromTuple where instead of string we need a bool followed by bytes

Reply:
Thanks. I tried a few different ways, but ultimately could not find a way to reuse much of the existing functions in eth_abi_utils.cc. I moved this logic to that file, but added two new functions, one for parsing the tuple, and one for parsing the list of tuples.  I did reuse ExtractBytesFromTuple and ExtractRow.